diff --git a/Business/ParabolicSAR.pm b/Business/ParabolicSAR.pm
index 80bbf7f..0b965b7 100644
--- a/Business/ParabolicSAR.pm
+++ b/Business/ParabolicSAR.pm
@@ -50,7 +50,7 @@ sub insert {
     my ($as,$am);
     croak "must set_alpha(as,am) before inserting data" unless defined( $am = $this->{am} ) and defined( $as = $this->{as} );
 
-    my ($y_low, $y_high) = @{$this->{y}};
+    my ($y_low, $y_high, $r) = @{$this->{y}};
     my ($open,$high,$low,$close);
 
     my $S;
@@ -58,6 +58,9 @@ sub insert {
     my $A = $this->{A};
     my $e = $this->{e};
 
+    my @_oe; # old $e
+    my @__r; # price range
+
     my $ls = $this->{ls};
 
     while( defined( my $ar = shift ) ) {
@@ -67,22 +70,21 @@ sub insert {
         # NOTE: we really only use open and close to initialize ...
         ($open,$high,$low,$close) = @$ar;
 
+        @__r = ($open<$close ? ($open,$close) : ($close,$open));
+
         if( defined $ls ) {
             # calculate sar_t
             # The Encyclopedia of Technical Market Indicators - Page 495
 
-            my @oe = @$e;
+            @_oe = @$e;
+
             $e->[HP] = $high if $high > $e->[HP]; # the highest point during the trend
             $e->[LP] = $low  if $low  < $e->[LP]; # the  lowest point during the trend
 
             if( $ls == LONG ) {
-                $S = $P + $A*($e->[HP] - $P); # adjusted upwards from the reset like so
-
-                # NOTE: many sources say you should flop short/long if you get
-                # inside the price range for the last *two* periods.  Amazon,
-                # Yahoo! and stockcharts dont' seem to do it that way.
+                $S = $P + $A*($_oe[HP] - $P); # adjusted upwards from the reset like so
 
-                if( $S > $low ) { # or $S > $y_low ) {
+                if( $S > $__r[0] or $S > $r->[0] ) { # greater than today's range-low or yesterday's
                     $ls = SHORT; # new short position
 
                     $S  = $e->[HP];
@@ -91,28 +93,32 @@ sub insert {
                     $e->[HP] = ($high>$y_high ? $high : $y_high);
                     $e->[LP] = ($low <$y_low  ? $low  : $y_low );
 
-                } elsif( $oe[HP] != $e->[HP] ) {
+                } elsif( $S > $low ) {
+                    $S = $low;
+
+                } elsif( $_oe[HP] != $e->[HP] ) {
                     $A += $as;
                     $A = $am if $A > $am;
                 }
 
             } else {
-                $S = $P + $A*($e->[LP] - $P); # adjusted downwards from the reset like so
-
-                # NOTE: many sources say you should flop short/long if you get
-                # inside the price range for the last *two* periods.  Amazon,
-                # Yahoo! and stockcharts dont' seem to do it that way.
+                $S = $P + $A*($_oe[LP] - $P); # adjusted downwards from the reset like so
 
-                if( $S < $high ) { # or $S < $y_high ) {
+                if( $S < $__r[1] or $S < $r->[1] ) {
                     $ls = LONG; # new long position
 
+                    warn "new LONG(@__r | @$r)($S)";
+
                     $S  = $e->[LP];
                     $A  = $as;
 
                     $e->[HP] = ($high>$y_high ? $high : $y_high);
                     $e->[LP] = ($low <$y_low  ? $low  : $y_low );
 
-                } elsif( $oe[LP] != $e->[LP] ) {
+                } elsif( $S < $high ) {
+                    $S = $high;
+
+                } elsif( $_oe[LP] != $e->[LP] ) {
                     $A += $as;
                     $A = $am if $A > $am;
                 }
@@ -150,7 +156,7 @@ sub insert {
     $this->{A}  = $A;
     $this->{ls} = $ls;
 
-    @{$this->{y}} = ($y_low, $y_high);
+    @{$this->{y}} = ($y_low, $y_high, \@__r);
 }
 
 sub start_with {
